core/src/main/scala/kafka/server/instrumentation/ProduceRequestInstrumentation.scala

```diff
@@ -0,0 +1,193 @@
+  var config: Config = extractConfigFromKafkaConfig(kafkaConfig)
+
+  override def reconfigurableConfigs: collection.Set[String] = ProduceRequestInstrumentationLogger.ReconfigurableConfigs
+
+  override def validateReconfiguration(newConfig: KafkaConfig): Unit = {
+    val thresholdMs = newConfig.longTailProduceRequestLogThresholdMs
+    if (thresholdMs < -1) {
+      throw new ConfigException(s"${KafkaConfig.LiLongTailProduceRequestLogThresholdMsProp} should be >= -1")
+    }
+    val ratio = newConfig.longTailProduceRequestLogRatio
+    if (ratio < 0 || ratio > 1) {
+      throw new ConfigException(s"${KafkaConfig.LiLongTailProduceRequestLogRatioProp} should be between 0 to 1")
+    }
+  }
+
+  /**
+   * Invoked when the associated DynamicBrokerConfig is changed
+   */
+  override def reconfigure(oldConfig: KafkaConfig, newConfig: KafkaConfig): Unit = {
+    config = extractConfigFromKafkaConfig(newConfig)
+  }
+
+  def extractConfigFromKafkaConfig(kafkaConfig: KafkaConfig): Config = Config(kafkaConfig.longTailProduceRequestLogThresholdMs,
+                                                                              kafkaConfig.longTailProduceRequestLogRatio)
+
+  /**
+   * Calculate the time in each stage by
+   * 1. Sort stages by timestamp
+   * 2. Make a sliding window of 2 on 1.
+   * 3. Diff the timestamp on the window to get the time spent on the stage of the first element in 2
+   */
+  def toTimeTakenInEachStageMessage(instrumentation: ProduceRequestInstrumentation): String = {
+    val marks = instrumentation.marks
+
+    // Enforce Finish is marked, so when sliding there is at list a full window
+    marks.get(Stage.Finish) match {
+      case None => instrumentation.markStage(Stage.Finish)
+      case _ =>
+    }
+
+    marks
+      .toSeq                          // Iterate over stages
+      .sortBy(s => (s._2, s._1.id))   // in order of marked timestamp then defined order of stage (to prevent multiple stages finish within 1ms)
+      .sliding(2)                     // Make sliding window to diff time taken in each stage
+      .map { window =>
+        val iter = window.iterator
+        val (stage1, timestamp1) = iter.next()
+        val (_, timestamp2) = iter.next()
+        // Use the next stage's timestamp mark minus this stage's timestamp mark to get time spent on current stage
+        // Last stage is on finish, no diff to check, i.e. no need to log
+        s"${stage1.toString}: ${timestamp2 - timestamp1}"
+      }
+      .mkString("(", ", ", ")")
+  }
+
+  /**
+   * Based on the config (e.g. probability funnel), logs the detail of the instrumentation
+   *
+   * @param request This is required to obtain the total time for threshold funnel
+   * @param instrumentation The instrumentation to log
+   */
+  def maybeLog(request: RequestChannel.Request,
+               instrumentation: ProduceRequestInstrumentation): Unit = {
+    val totalTimeMs = nanosToMs(time.nanoseconds() - request.startTimeNanos)
```

Comment:
It would be better to use the time of Stage.Finish as the end time to reduce inconsistency.

Reply:
Change the implementation to use init & finish.

Reply:
Actually I changed to use the request metrics (as used in updated metrics method).